File: preschool-management-backend/src/main/java/com/vusystem/preschool_management_backend/modules/core/services/MealRegistrationService.java
package com.vusystem.preschool_management_backend.modules.core.services;


import com.vusystem.preschool_management_backend.modules.core.dto.request.MonthlyMealRegistrationRequest;
import com.vusystem.preschool_management_backend.modules.core.dto.response.ChildMonthlyMealStatsResponse;
import com.vusystem.preschool_management_backend.modules.core.dto.response.MealRegistrationResponse;
import com.vusystem.preschool_management_backend.modules.core.dto.response.MealStatisticsResponse;

import java.time.LocalDate;
import java.util.List;

public interface MealRegistrationService {


    // xử lý tự động đăng ký hoặc cắt cơm hàng loạt theo tháng để hỗ trợ tác vụ hành chính
    void processMonthlyRegistration(MonthlyMealRegistrationRequest request);

    void processDailyRegistration(com.vusystem.preschool_management_backend.modules.core.dto.request.DailyMealRegistrationRequest request);

    // cho phép giáo viên/admin đăng ký ngoại lệ vượt qua ràng buộc thời gian
    void overrideDailyRegistration(com.vusystem.preschool_management_backend.modules.core.dto.request.DailyMealRegistrationRequest request);

    List<MealRegistrationResponse> getRegistrationsByClassAndDate(Long classId, LocalDate date);

    List<MealRegistrationResponse> getRegistrationsByChildAndDateRange(Long childId, LocalDate startDate, LocalDate endDate);

    List<ChildMonthlyMealStatsResponse> getMonthlyMealStatsByClass(Long classId, int month, int year);
    
    MealStatisticsResponse getMealStatistics(LocalDate startDate, LocalDate endDate);

    // tự động đồng bộ hóa việc cắt cơm khi học sinh có đơn xin nghỉ được duyệt
    void cancelMealsForLeave(Long childId, LocalDate startDate, LocalDate endDate);

    // tự động khôi phục suất ăn khi phụ huynh hủy đơn xin nghỉ đã duyệt
    void restoreMealsForLeaveCancel(Long childId, LocalDate startDate, LocalDate endDate);
}

File: mobile/src/services/mealRegistration.service.ts
import axiosInstance from '../config/api.client';

export interface MonthlyMealRegistrationRequest {
  childId: number;
  month: number;
  year: number;
  isRegistered: boolean;
  mealTypes: string[];
}

export interface DailyMealRegistrationRequest {
  childId: number;
  date: string;
  isRegistered: boolean;
  mealTypes: string[];
}

export interface MealRegistrationResponse {
  id: number;
  childId: number;
  childFullName: string;
  className: string;
  date: string;
  mealType: string;
  status: string;
}

export const mealRegistrationService = {
  processMonthlyRegistration: async (request: MonthlyMealRegistrationRequest): Promise<void> => {
    await axiosInstance.post(`/core/meal-registrations/monthly`, request);
  },

  processDailyRegistration: async (request: DailyMealRegistrationRequest): Promise<void> => {
    await axiosInstance.post(`/core/meal-registrations/daily`, request);
  },

  getRegistrationsByChildAndDateRange: async (childId: number, startDate: string, endDate: string): Promise<MealRegistrationResponse[]> => {
    const response = await axiosInstance.get(`/core/meal-registrations/child/${childId}`, {
      params: { startDate, endDate }
    });
    return response.data.data;
  }
};
